# 01. Importing libraries

In [3]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data
#Import "products.csv" data from "Original Data" folder as df_products
& "orders_wrangled.csv" from "prepared data" folder as df_ords

In [9]:
df = pd.read_csv(r'C:\Users\alexl\Downloads\4.3_orders_products (1)\orders.csv', index_col=False)

In [19]:
path =r'C:\Users\alexl\CareerFoundry\Instacart Basket Analysis'

In [21]:
df_products=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

In [24]:
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col=False)

In [116]:
#Create a dataframe

df_test=pd.DataFrame()

In [117]:
#Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [118]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [119]:
df_products.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [120]:
df_nan=df_products[df_products['product_name'].isnull()==True]

In [121]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [122]:
df_products.shape

(49693, 5)

In [123]:
df_products_clean=df_products[df_products['product_name'].isnull()==False]

In [124]:
df_products_clean.shape

(49677, 5)

In [125]:
df_dups=df_products_clean[df_products_clean.duplicated()]

In [126]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [127]:
df_products_clean.shape

(49677, 5)

In [128]:
df_products_clean_no_dups=df_products_clean.drop_duplicates()

In [129]:
df_products_clean_no_dups.shape

(49672, 5)

# 03. Assignment

In [130]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Min orders_day_of_week may indicate 

In [131]:
#Check for mixed-type data in your df_ords dataframe. 
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [132]:
print(col)

days_since_prior_order


In [133]:
#Fix mixed-type data. Column in question should contain observations of a single data type.
df_test['mix'] = df_test['mix'].astype('str')

In [134]:
#Check for missing values in your df_ords dataframe. 
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

206209 missing values for days_since_prior_order column

In [135]:
df_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [136]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

Leave NaN as is. Indication that this is the customer's first order, and therefore no prior days since last order. 

In [139]:
#Run a check for duplicate values in your df_ords data.

In [140]:
df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [141]:
df_ords_clean.shape

(3214874, 8)

In [142]:
#Address the duplicates using an appropriate method. 
df_ords_clean_no_dups=df_ords_clean.drop_duplicates()

In [143]:
df_ords_clean_no_dups.shape

(3214874, 8)

No changes necessary

In [30]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [31]:
df_products.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))